### ISSUE TITLE
Feature request: prefer module over main

### ISSUE THREAD

**Issue by rtransat** [codesandbox/codesandbox-client#3305]
# 🐛 bug report

## Description of the problem

I've developed a npm package with [vue-sfc-rollup](https://github.com/team-innovation/vue-sfc-rollup) and the build in codesandbox use the ssr file not the esm.
You can see the [package.json](https://github.com/rtransat/vhue/blob/master/package.json#L22-L23) of my project.

It's works with a vue cli project in local.

## How has this issue affected you? What are you trying to accomplish?

It seems that codesandbox use the "main" entry point and I have the following error:

```
[Vue warn]: Error in render: "TypeError: _vm._ssrNode is not a function"
__vue_render__@https://zq0n5.csb.app/node_modules/vhue/dist/vhue.ssr.js:4440:66
```

### Your Environment

| Software         | Name/Version |
| ---------------- | ------------ |
| Сodesandbox      | Online
| Browser          | Firefox
| Operating System | Linux Mint 18.3

**Comment by mgdodge** [comment]
For completeness' sake, the vue-sfc-rollup is based on [this document](https://vuejs.org/v2/cookbook/packaging-sfc-for-npm.html) from the vue cookbook, along with updates to that document currently pending in [this pull request](https://github.com/vuejs/vuejs.org/pull/2172).

In short, the package.json has a modern es build referenced with the `module` field, since all modern build tools (webpack, rollup, etc.) recognize that. The `main` field is used for the ssr-optimized build since that is the field that node.js sees by default, as well as nuxt.js and other Vue ssr utilities.

If codesandbox follows either webpack or rollup methodology, it should be picking up the `modern` build, not the `main` build.

**Comment by lbogdan** [comment]
I kind of fixed this in #3319 , you can test the linked sandbox here: https://pr3319.build.csb.dev/s/zq0n5 . Unfortunately the change breaks other sandboxes, and also [impacts performance](https://github.com/codesandbox/codesandbox-client/pull/3319#pullrequestreview-341588027).

Would it be possible to set the `browser` field in `package.json` instead, which is also the [`webpack` first preferred](https://webpack.js.org/configuration/resolve/#resolvemainfields) in web builds?

**Comment by mgdodge** [comment]
### TL;DR
Adding the `browser` field will fix the specific Vue issue mentioned here, but only if #3319 contains the `browser` field. However, the `module` field, as a widely-used de-facto standard, should also be supported.

### Details

The webpack preferred you link to shows `browser` first and `module` second - so according to this logic, #3319 is a step in the right direction as far as maintaining compatibility with the rest of the ecosystem. This applies across the board, for all npm modules, not just Vue-based ones.

[As mentioned previously](https://github.com/codesandbox/codesandbox-client/issues/3305#issuecomment-573378725), in the Vue ecosystem, webpack is typically configured slightly differently from default in an SSR context. For example, nuxt.js (arguably the most widely-used server-side-rendering platform for Vue) [actively ignores the `browser` field](https://github.com/nuxt/nuxt.js/blob/dev/packages/webpack/src/config/server.js#L48), looking instead for `main` and then `module`. Quasar works the same way, as would likely all other webpack/node ssr usages. But in a client-side context, it seems the webpack defaults are used.

In any case, by adding the `browser` field to this package or any similarly packaged module, it would have to point to the esm build (essentially _skipping over_ the `main` field in codesandbox, and _duplicating_ `module` for all other bundlers.)

Here's a breakdown of some tooling tests I ran earlier this morning:

**Leaving `browser` out of package.json:**
- Webpack gets esm build from `module`
- nuxtjs/quasar will use ssr from `main` in server job, esm from `module` in client job
- express + vue server renderer will use ssr from `main`
- codesandbox breaks trying to use ssr from `main`
  - Implementing `module` via #3319 would fix this, codesandbox would use esm from `module`

**Setting package.json `browser` to a esm build:**
- Webpack gets esm build from `browser`
- nuxtjs/quasar will use ssr from `main` in server job, esm build from `browser` in client job
- express + vue server renderer will use ssr from `main`
- codesandbox breaks trying to use ssr from `main`
  - Implementing `browser` via #3319 would fix this, codesandbox would use esm from `browser`
- the `module` field is no longer referenced at all

It feels a little dirty to duplicate the already-present `module` field just to get around the default codesandbox behavior, but it *is* an actual standard field, and it does address the issue. It is worth noting, however, that **rollup.js will see the `module` field by default, but won't see the `browser` field without a plugin.** I can't speak to every bundler scenario, but it is reasonable to assume that others may be looking for the `module` field as well, and numerous libraries will be providing that in lieu of `browser`. As a de-facto standard, should remain in place and receive support, even if it is duplication.

What it all boils down to is this: adding the `browser` field is an acceptable fix, but only if #3319 contains the `browser` field. I also believe it should contain the `module` field to address instances where the `browser` field is not present, as is likely in much of the ecosystem. This will keep codesandbox compatible with everyone else's expectations.

I would also happily update my utility to properly configure the `browser` field, duplication aside, as well as the vue cookbook PR I currently have pending.

**Comment by lbogdan** [comment]
@mgdodge We're already honoring the `browser` field, even without #3319 , as we use https://github.com/CompuIves/node-browser-resolve for resolving.

**Comment by lbogdan** [comment]
@mgdodge Hm, so you're saying that it's not working in CodeSandbox, even with a `browser` field? Do you have a sandbox I can look at?

**Comment by mgdodge** [comment]
I didn't actually have one to test, I looked at #3319 and saw someone asking browser to be added. I assumed it wasn't there yet (like how nuxt.js  does webpack.) Also saw way back at [PR #853](https://github.com/codesandbox/codesandbox-client/pull/853) that the `browser` field was being removed, again assumed it was a global thing. Guess it's not.

I personally don't have any sandboxes yet. Gonna have to publish a couple of throw-away npm modules to do some comparison.

**Comment by lbogdan** [comment]
> saw someone asking browser to be added.

#3319 is a very naive fix, and as it is now, it prefers `module` with the highest priority. It was me, what I was saying was that maybe we should keep `browser` as the first preferred.

>Gonna have to publish a couple of throw-away npm modules to do some comparison.
Maybe take a look at [CodeSandbox CI](https://codesandbox.io/docs/ci). 🙂

**Comment by mgdodge** [comment]
Okay, just tried things in a new Vue cli sandbox, didn't have to create throw-away npm modules either. Just imported from the directory instead, and included the package.json for the import to resolve.

https://codesandbox.io/s/compassionate-rain-p8l6g

Looks like the `browser` field does in fact work with the esm build in there. 👍 

Now, just to be sure, I created one based on nuxt

https://codesandbox.io/s/keen-kalam-s1sfv

I had a little more trouble getting nuxt to refresh, but it seems to be working there as well, from what I can tell. I can't see the ssr output, but it should be fine since (as my earlier tests indicate) the ssr builds never factored in the `browser` field.

So all indications are that `browser` set to the esm build seems to be doing the trick in codesandbox. @rtransat can manually add that to his package.json file and his sandbox should start working. I will update vue-sfc-rollup to set that field as well.

Do you think that adding proper `module` field support is still on the radar, or does this make it a non-starter? It doesn't directly affect me, more just curious how codesandbox will handle/document this inconsistency.

**Comment by lbogdan** [comment]
Nuxt sandboxes are server / container sandboxes, they run inside an actual container pretty much the same as in a local development environment, using `webpack` & co., so they are not at all affected by the in-browser *webpack-like* bundler.

**Comment by CompuIves** [comment]
Interesting. Another detail: if there's no `main` field we'll default to `module`. I'm kind of hoping we can include tree shaking in the bundler before we default to `module`, as it's much slower. An example of this, if we'd include `lodash-es` and don't have treeshaking, we would have to download over 600 files before we can bundle.
